#!/bin/bash

### rename_artificial_reads ###

# This script renames each read generated by artfastqgen in separate jobs from many
# chromosome fragment files to make it unique so the reads can be combined from each
# output file into one final output file

projectname="benchmark"
# specify length of reads and coverage for inDir path:
lengths=( 700 900 2000 5000 7500 )

# specify the number of lines chromosome had been split into:
split_no="16390_lines"

#make directory hierachy:
homeDir="/home/jamtor"
binDir="$homeDir/local/bin"
projectDir="$homeDir/projects/$projectname"
resultsDir="$projectDir/results"

for length in ${lengths[@]}; do

	echo -e
	echo "Reads of the following length are to be processed: $length"

	#input/output directories:
	inDir="$resultsDir/long.artfastqgen.reads/from_$split_no"_chunks/"$length"bp
	outDir="$inDir/to_combine"
	
	mkdir -p $outDir
	
	#scripts/log directories
	scriptsPath="$projectDir/scripts/readgen"
	logDir="$scriptsPath/logs"
	
	mkdir -p $logDir
	
	echo -e
	echo "The inDir is:"
	echo $inDir
	echo -e
	echo "The outDir is:"
	echo $outDir
	echo -e
	echo "The logDir is:"
	echo $logDir
	
	# make one directory for reads of each length and move the
	# corresponding read files to that directory:
	#mkdir $length
	
	#for file in *$length*; do
	#	echo $file
	#	mv $file $length
	#done;
	
	filesTotal=( $(ls $inDir/*.fastq) )

	j=0
	i=0

	while [ $j -lt ${#filesTotal[@]} ]; do
		inFile1=${filesTotal[$j]}
		inFile2=${filesTotal[$j+1]}
		uniqueID1=`basename $inFile1 | sed 's/.fastq//'`
		uniqueID2=`basename $inFile2 | sed 's/.fastq//'`

		echo -e
		echo $j
		echo $i

		echo -e
		echo "This is the first uniqueID:"
		echo $uniqueID1
		echo -e
		echo "This is the second uniqueID:"
		echo $uniqueID2

		newname1="$uniqueID1.renamed.fastq"
		newname2="$uniqueID2.renamed.fastq"

		echo -e
		echo "The first renamed file is:"
		echo $newname1
		echo -e
		echo "The second renamed file is:"
		echo $newname2

		more $inFile1 | sed "s/\@HWI/\@HWI`echo $i`/g" | sed 's/\+.*/\+/' > "$outDir/$newname1"
		more $inFile2 | sed "s/\@HWI/\@HWI`echo $i`/g" | sed 's/\+.*/\+/' > "$outDir/$newname2"

		i=$(($i+1))
		j=$(($j+2))

	done;
done;



#	files="$inDir/*.fastq"
#	i=0
#	for file in $files; do
#		uniqueID1=`basename $file | sed 's/.'
#		echo -e
#		echo "The file used is:"
#		echo $file
#		newname=`basename $file | sed 's/fastq/renamed.fastq/g'`
#		echo -e
#		echo "The renamed file is:"
#		echo $newname
#	
#		#grep @HWI $file > "$outDir/count.txt"
#		#count=`wc -l "$outDir/count.txt"`
#		#count=`echo $count | awk -F ' ' '{ print $1 }'`
#	
#		#in order to rename individual reads we insert
#		#incremental numbers after the HWI in the ID
#		#this is followed by the removal of anything after the + sign
#		#in the third line of the fastq output, where artfastqgen places 
#		#another ID for some reason. :/ 
#
#		more $file | sed "s/\@HWI/\@HWI`echo $i`/g" | sed 's/\+.*/\+/' > "$outDir/$newname"
#		
#		#more $file | sed "s/\@HWI/\@HWI$i/g" | sed 's/ @HWI-/-/g'` > "$outDir/temp.txt"
#		#sed 's/@/\n&/g' "$outDir/temp.txt" > "$outDir/temp2.txt"
#		#sed 's/+/\n&/g' "$outDir/temp2.txt" > "$outDir/temp.txt"
#		#sed 's/1 /&\n/g' "$outDir/temp.txt" > "$outDir/temp2.txt"
#		#sed 's/\+.*/\+/' "$outDir/temp2.txt" > "$outDir/$newname"
#		i=$(($i+1))
	#rm "$outDir/temp*"
