#!/bin/bash

### join_artificial_reads ###

# This script joins the fragmented files generated by
# artfastqgen together into one final output file and
# organises them into fasta format

projectname="benchmark"
# specify length of reads for inDir path:
lengths=( 50 75 120 200 300 400 700 900 2000 5000 7500 )

# specify the number of lines chromosome had been split into:
split_no="16390_lines"

#make directory hierachy:
homeDir="/home/jamtor"
binDir="$homeDir/local/bin"
projectDir="$homeDir/projects/$projectname"
resultsDir="$projectDir/results"

for length in ${lengths[@]}; do

	echo -e
	echo "Reads of the following length are to be processed: $length"
	
	#input/output directories:
	readsDir="$resultsDir/long.artfastqgen.reads/from_$split_no"_chunks/"$length"bp
	inDir="$readsDir/to_combine"
	tempDir="$readsDir/combined"
	outDir="$readsDir/final"

	mkdir -p $tempDir
	mkdir -p $outDir

	outExt=".combined.fastq"

	#scripts/log directories
	scriptsPath="$projectDir/scripts/readgen"
	logDir="$scriptsPath/logs"
	
	mkdir -p $logDir

	echo -e
	echo "The inDir is:"
	echo $inDir
	echo -e
	echo "The tempDir is:"
	echo $tempDir
	echo -e
	echo "The outDir is:"
	echo $outDir
	echo -e
	echo "The logDir is:"
	echo $logDir
	
	filesTotal=( $(ls $inDir/*.renamed.fastq) )
	
	#unset filesTotal[${#filesTotal[@]}-1]
	#unset filesTotal[${#filesTotal[@]}-1]

	j=0
	echo -e
	echo "Total files are:"
	echo ${#filesTotal[@]}


	#control for appending to a file
	rm $tempDir/$length.1.combined.fastq
	rm $tempDir/$length.2.combined.fastq

	while [ $j -lt ${#filesTotal[@]} ]; do
		inFile1=${filesTotal[$j]}
		inFile2=${filesTotal[$j+1]}
		uniqueID1=`basename $inFile1 | sed 's/.renamed.fastq//'`
		uniqueID2=`basename $inFile2 | sed 's/.renamed.fastq//'`

		echo -e
		echo "This is the first uniqueID:"
		echo $uniqueID1
		echo -e
		echo "This is the second uniqueID:"
		echo $uniqueID2
		combine_line1="cat $inFile1 >> $tempDir/$length.1.combined.fastq"
		
		#cat $inFile1 >> $tempDir/$length.1.combined.fastq

		echo -e
		echo "This is the combine_line1:"
		echo $combine_line1

		# define the previous job number as value of j for the previous qsub to hold job until last
		# job is done: 
		last_job_no=$(($j-2))

		echo -e
		echo "Last job no: "$last_job_no
		echo -e
		echo "Current job no: "$j

		qsub -N one$length.$j -hold_jid one$length.$last_job_no -wd $logDir -b y -j y -R y -pe smp 1 $combine_line1

		combine_line2="cat $inFile2 >> $tempDir/$length.2.combined.fastq"
		echo -e
		echo "This is the combine_line2:"
		echo $combine_line2

		#cat $inFile2 >> $tempDir/$length.2.combined.fastq

		qsub -N two$length.$j -hold_jid two$length.$last_job_no -wd $logDir -b y -j y -R y -pe smp 1 $combine_line2

#		if [ -f $tempDir/$length.1.combined.fastq ]; then
#			echo -e
#			echo "First initial combined file exists"
#		else
#			echo -e
#			echo "Creating first initial combined file"
#			reads1=`cat $inFile1`
#			echo $reads1 >> "$tempDir/$length.1.combined.fastq"
#		fi
#
#		if [ -f $tempDir/$length.2.combined.fastq ]; then
#			echo -e
#			echo "Second initial combined file exists"
#		else
#			echo -e
#			echo "Creating second initial combined file"
#			reads2=`cat $inFile2`
#			echo $reads2 >> "$tempDir/$length.2.combined.fastq"
#		fi

		files=( $length.1.combined.fastq $length.2.combined.fastq )
		for file in ${files[@]}; do

			uniqueID=`basename $file | sed 's/.combined.fastq//'`
			outFile="$outDir/$uniqueID$outExt"

			echo -e
			echo "This is the outfile:"
			echo $outFile
			
#			sed 's/1 /&\n/g' $file > "$tempDir/temp.txt"
#			sed 's/ //g' "$tempDir/temp.txt" > "$tempDir/temp2.txt"
#			sed 's/@/\n&/g' "$tempDir/temp2.txt" > "$tempDir/temp.txt"
#			sed 's/+/\n&/g' "$tempDir/temp.txt" > "$tempDir/temp2.txt"
#			sed 's/+/&\n/g' "$tempDir/temp2.txt" > "$tempDir/temp.txt"
#			sed '/./,$!d' "$tempDir/temp.txt" > "$tempDir/temp2.txt"
#			more "$tempDir/temp2.txt" | perl -ne '$line=$_;if(length($line)>121){@array=$line=~/\w{1,121}/g;print(join("\n", @array));print "\n"}else{print $line}' > $outFile
			#rm $outDir/temp*
		done;
		j=$(($j+2))
	done;
	
#	inFile1="$inDir/chr21bv*bp.1.renamed.fastq"
#	inFile2="$inDir/chr21bv*bp.2.renamed.fastq"
#	uniqueID1=`basename $inFile1 | sed 's/.renamed.fastq//'`
#	uniqueID2=`basename $inFile2 | sed 's/.renamed.fastq//'`
#
#	echo -e
#	echo "This is the first uniqueID:"
#	echo $uniqueID1
#	echo -e
#	echo "This is the second uniqueID:"
#	echo $uniqueID2
#	combine_line1="cat $inFile1 >> $tempDir/$length.1.combined.fastq"
#	
#	#cat $inFile1 >> $tempDir/$length.1.combined.fastq
#	echo -e
#	echo "This is the combine_line1:"
#	echo $combine_line1
#	qsub -N $uniqueID1 -wd $logDir -b y -j y -R y -pe smp 1 $combine_line1
#	combine_line2="cat $inFile2 >> $tempDir/$length.2.combined.fastq"
#	echo -e
#	echo "This is the combine_line2:"
#	echo $combine_line2
#	#cat $inFile2 >> $tempDir/$length.2.combined.fastq
#	qsub -N $uniqueID2 -wd $logDir -b y -j y -R y -pe smp 1 $combine_line2

done;