#!/bin/bash

### rename_artificial_reads ###

# This script renames each read generated by artfastqgen in separate jobs from many
# chromosome fragment files to make it unique so the reads can be combined from each
# output file into one final output file

projectname="benchmark"
# specify length of reads and coverage for inDir path:
lengths=( "2000bp30x" )

#make directory hierachy:
homeDir="/home/jamtor"
binDir="$homeDir/local/bin"
projectDir="$homeDir/projects/$projectname"
resultsDir="$projectDir/results"

for length in ${lengths[@]}; do

	echo -e
	echo "Reads of the following length are to be processed: $length"

	#input/output directories:
	inDir="$resultsDir/long.artfastqgen.reads/$length"
	outDir="$inDir/to_combine"
	
	#scripts/log directories
	scriptsPath="$projectDir/scripts/readgen"
	logDir="$scriptsPath/logs"
	
	mkdir -p $outDir
	mkdir -p $logDir
	
	echo -e
	echo "The inDir is:"
	echo $inDir
	echo -e
	echo "The outDir is:"
	echo $outDir
	echo -e
	echo "The logDir is:"
	echo $logDir
	
	# make one directory for reads of each length and move the
	# corresponding read files to that directory:
	#mkdir $length
	
	#for file in *$length*; do
	#	echo $file
	#	mv $file $length
	#done;
	
	files="$inDir/*.1.fastq"
	i=0
	for file in $files; do
		echo -e
		echo "The file used is:"
		echo $file
		newname=`basename $file | sed 's/1.fastq/renamed.fastq/g'`
		echo -e
		echo $newname
	
		#grep @HWI $file > "$outDir/count.txt"
		#count=`wc -l "$outDir/count.txt"`
		#count=`echo $count | awk -F ' ' '{ print $1 }'`
	
		reads=`sed "s/@HWI/@HWI$i/g" $file`
		reads=`echo $reads | sed 's/ @HWI-/-/g'`
		echo $reads > "$outDir/temp.txt"
		sed 's/@/\n&/g' "$outDir/temp.txt" > "$outDir/temp2.txt"
		sed 's/+/\n&/g' "$outDir/temp2.txt" > "$outDir/temp.txt"
		sed 's/1 /&\n/g' "$outDir/temp.txt" > "$outDir/temp2.txt"
		sed 's/\+.*/\+/' "$outDir/temp2.txt" > "$outDir/$newname"
		i=$(($i+1))
	done;
	#rm "$outDir/temp*"
done;