#!/bin/bash

### join_artificial_reads ###

# This script joins the fragmented files generated by
# artfastqgen together into one final output file and
# organises them into fasta format

projectname="benchmark"
# specify length of reads for inDir path:
lengths=( 50 75 120 200 300 400 700 900 2000 5000 7500 )

# specify the number of lines chromosome had been split into:
split_no="16390_lines"

#make directory hierachy:
homeDir="/home/jamtor"
projectDir="$homeDir/projects/$projectname"
resultsDir="$projectDir/results"

for length in ${lengths[@]}; do

	echo -e
	echo "Reads of the following length are to be processed: $length"
	
	#input/output directories:
	readsDir="$resultsDir/long.artfastqgen.reads/from_$split_no"_chunks/"$length"bp
	inDir="$readsDir/combined"
	outDir="$readsDir/final"

	mkdir -p $outDir

	outExt=".fastq"

	#scripts/log directories
	scriptsPath="$projectDir/scripts/readgen"
	logDir="$scriptsPath/logs"
	
	mkdir -p $logDir

	echo -e
	echo "The inDir is:"
	echo $inDir
	echo -e
	echo "The outDir is:"
	echo $outDir
	echo -e
	echo "The logDir is:"
	echo $logDir

	inFiles=( $(ls $inDir/*combined.fastq) )
	for inFile in ${inFiles[@]}; do

		uniqueID=`basename $inFile | sed 's/.combined.fastq//'`
		outFile="$outDir/$uniqueID$outExt"

		echo -e
		echo "This is the inFile:"
		echo $inFile
		echo -e
		echo "this is the outFile:"
		echo $outFile


		f1="sed 's/1 /&\n/g' $inFile > "$inDir/temp.txt""
		echo -e
		echo $f1

		# remove the spaces between all reads/ids:
		f2="sed 's/ //g' "$inDir/temp.txt" > "$inDir/temp2.txt""
		echo -e
		echo $f2
		f3="sed 's/@/\n&/g' "$inDir/temp2.txt" > "$inDir/temp.txt""
		echo -e
		echo $f3
		f4="sed 's/+/\n&/g' "$inDir/temp.txt" > "$inDir/temp2.txt""
		echo -e
		echo $f4
		f5="sed 's/+/&\n/g' "$inDir/temp2.txt" > "$inDir/temp.txt""
		echo -e
		echo $f5
		f6="sed '/./,$!d' "$inDir/temp.txt" > "$inDir/temp2.txt""
		echo -e
		echo $f6
		f7="more "$inDir/temp2.txt" | perl -ne '$line=$_;if(length($line)>121){@array=$line=~/\w{1,121}/g;print(join("\n", @array));print "\n"}else{print $line}' > $outFile"
		echo -e
		echo $f7

#		qsub -N $uniqueIDf1 -wd $logDir -b y -j y -R y -pe smp 1 $f1
#		qsub -N $uniqueIDf2 -wd $logDir -b y -j y -R y -pe smp 1 $f2
#		qsub -N $uniqueIDf3 -wd $logDir -b y -j y -R y -pe smp 1 $f3
#		qsub -N $uniqueIDf4 -wd $logDir -b y -j y -R y -pe smp 1 $f4
#		qsub -N $uniqueIDf5 -wd $logDir -b y -j y -R y -pe smp 1 $f5
#		qsub -N $uniqueIDf6 -wd $logDir -b y -j y -R y -pe smp 1 $f6
#		qsub -N $uniqueIDf7 -wd $logDir -b y -j y -R y -pe smp 1 $f7

		#rm $outDir/temp*
	done;
done;



